src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/FederationAccessManagerWithCache.java

```diff
@@ -0,0 +1,98 @@
+
+import se.liu.ida.hefquin.engine.federation.BRTPFServer;
+import se.liu.ida.hefquin.engine.federation.Neo4jServer;
+import se.liu.ida.hefquin.engine.federation.SPARQLEndpoint;
+import se.liu.ida.hefquin.engine.federation.TPFServer;
+import se.liu.ida.hefquin.engine.federation.access.BRTPFRequest;
+import se.liu.ida.hefquin.engine.federation.access.CardinalityResponse;
+import se.liu.ida.hefquin.engine.federation.access.FederationAccessException;
+import se.liu.ida.hefquin.engine.federation.access.FederationAccessManager;
+import se.liu.ida.hefquin.engine.federation.access.Neo4jRequest;
+import se.liu.ida.hefquin.engine.federation.access.SPARQLRequest;
+import se.liu.ida.hefquin.engine.federation.access.SolMapsResponse;
+import se.liu.ida.hefquin.engine.federation.access.StringResponse;
+import se.liu.ida.hefquin.engine.federation.access.TPFRequest;
+import se.liu.ida.hefquin.engine.federation.access.TPFResponse;
+import se.liu.ida.hefquin.engine.query.TriplePattern;
+
+public class FederationAccessManagerWithCache implements FederationAccessManager 
+{
+	private FederationAccessManager fedAccMan;
+	private Map<TriplePattern, CompletableFuture<CardinalityResponse>> cacheMap;
+	
+	public FederationAccessManagerWithCache(FederationAccessManager fedAccMan) {
+		this.fedAccMan = fedAccMan;
+		this.cacheMap = new HashMap<TriplePattern, CompletableFuture<CardinalityResponse>>();
+	}
+	
+	protected void addResultToCache(TriplePattern tp, CompletableFuture<CardinalityResponse> response) {
+		cacheMap.put(tp, response);
+	}
+	
+	@Override
+	public CompletableFuture<SolMapsResponse> issueRequest(SPARQLRequest req, SPARQLEndpoint fm)
+			throws FederationAccessException {
+		return fedAccMan.issueRequest(req, fm);
+	}
+
+	@Override
+	public CompletableFuture<TPFResponse> issueRequest(TPFRequest req, TPFServer fm) throws FederationAccessException {
+		return fedAccMan.issueRequest(req, fm);
+	}
+
+	@Override
+	public CompletableFuture<TPFResponse> issueRequest(TPFRequest req, BRTPFServer fm)
+			throws FederationAccessException {
+		return fedAccMan.issueRequest(req, fm);
+	}
+
+	@Override
+	public CompletableFuture<TPFResponse> issueRequest(BRTPFRequest req, BRTPFServer fm)
+			throws FederationAccessException {
+		return fedAccMan.issueRequest(req, fm);
+	}
+
+	@Override
+	public CompletableFuture<StringResponse> issueRequest(Neo4jRequest req, Neo4jServer fm)
+			throws FederationAccessException {
+		return fedAccMan.issueRequest(req, fm);
+	}
+
+	@Override
+	public CompletableFuture<CardinalityResponse> issueCardinalityRequest(SPARQLRequest req, SPARQLEndpoint fm)
+			throws FederationAccessException {
+		return fedAccMan.issueCardinalityRequest(req, fm);
+	}
+
+	@Override
+	public CompletableFuture<CardinalityResponse> issueCardinalityRequest(TPFRequest req, TPFServer fm)
+			throws FederationAccessException {
+		System.out.println(cacheMap.values());
+		CompletableFuture<CardinalityResponse> cacheResponse = cacheMap.get(req.getQueryPattern());
```

Comment:
What I mentioned about `final` also holds for local variables such as this one. Unless a local variable is explicitly meant to be reassigned within the method, it should be marked `final` as a safety measure.
```suggestion
		final CompletableFuture<CardinalityResponse> cacheResponse = cacheMap.get(req.getQueryPattern());
```